core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java

```diff
@@ -178,38 +185,45 @@ public void start() {
    @Override
    @Stop
    public void stop() {
-      // If needed, clear the persistent store before stopping
-      if (clearOnStop)
-         clearAllStores(AccessMode.BOTH);
-
-      Set undelegated = new HashSet();
-      Consumer<CacheWriter> stopWriters = writer -> {
-         writer.stop();
-         if (writer instanceof DelegatingCacheWriter) {
-            CacheWriter actual = undelegate(writer);
-            actual.stop();
-            undelegated.add(actual);
-         } else {
-            undelegated.add(writer);
+      storesMutex.writeLock().lock();
+      try {
+         // If needed, clear the persistent store before stopping
+         if (clearOnStop) {
+            clearAllStores(AccessMode.BOTH);
```

Comment:
Unfortunately this method acquires the read lock which will deadlock. Maybe move the write lock acquisition to just after?

Reply:
deadlock where? you can acquire the read-lock with the write-lock acquired.

Reply:
Whoops mixed up order. Disregard.